criu/files.c

```diff
@@ -1568,12 +1568,25 @@ int inherit_fd_add(int fd, char *key)
 	if (fd > inh_fd_max)
 		inh_fd_max = fd;
 
-	inh->inh_id = key;
+	inh->inh_id = xstrdup(key);
+	if (inh->inh_id == NULL)
+		return -1;
+
 	inh->inh_fd = fd;
 	list_add_tail(&inh->inh_list, &opts.inherit_fds);
 	return 0;
 }
 
+void inherit_fd_clean(void)
+{
+	struct inherit_fd *inh;
+
+	list_for_each_entry(inh, &opts.inherit_fds, inh_list) {
+		xfree(inh->inh_id);
+		inh->inh_id = NULL;
```

Comment:
xfree(inh) and you will need to use list_for_each_entry_safe.

I think we don't need inherit_fd_clean, CRIU usually exits right after restoring processes.

Reply:
> I think we don't need inherit_fd_clean, CRIU usually exits right after restoring processes.

Agree, I've removed it. What's left of the patch is more or less trivial.